File: src/json.cpp
#include <stdint.h>
#include <string>
#include <vector>
#include <json.hpp>
#include <base64.hpp>

using json = nlohmann::json;
using namespace std; 


// #define MAX_IP_LEN 16
// #define MESSAGE_TYPE_LEN 30
// #define MAX_CLIENTS 100
// #define MAX_KEY_LEN 256
// #define MAX_DATA_LEN 1024
// #define MAX_SIGNATURE_LEN 256
// #define MAX_SERVERS 15
// #define MAX_PARTICIPANTS 30
// #define MAX_FINGERPRINTS_LEN 64
// #define MAX_DESTINATIONS 15
// #define MAX_CHAT_MSG_LEN 1024
// #define MAX_AES_IV_LEN 16
// #define MAX_ENCRYPTED_KEY_LEN 256

static uint32_t m_counter = 0;

struct ClientMessageStructure {
    string type;
    string data;
    uint32_t counter;
    string signature;
};

struct ClientHelloData{
    string type;
    string public_key;
};

struct ClientHelloMessage {
    ClientHelloData data;
};

struct ServerClientListRequest {
    string type;
};

struct ServerInfo {
    string address;
    vector<string> clients;
};

struct ClientList {
    string type;
    vector<ServerInfo> servers;
};

struct ServerHelloData {
    string type;
    string sender;
};

struct  ServerHelloMessage {
    ServerHelloData data;
};

struct ServerClientUpdateRequest {
    string type;
};

struct ServerClientUpdate{
    string type;
    vector<string> clients;
    int client_count;
};

struct ChatMessage {
    string participants;
    string message;
};

struct ChatDataStructure {
    string type;
    vector<string> destinations;
    string iv;
    vector<string> symm_keys;
    string chat;
};

json createSignedMessage(const json& data) {
    json message;
    message["type"] = "signed_data";
    message["data"] = data;
    message["counter"] = m_counter;

    string messageStr = message.dump();
    string signatureInput = messageStr + to_string(m_counter);
    string signature = macaron::Base64::Encode(signatureInput);

    message["signature"] = signature;
    m_counter++;
    return message;
}


json helloMessage(const string& publicKey) {

    json helloData;
    helloData["type"] = "hello";
    helloData["public_key"] = publicKey; // still needs implementation

    return createSignedMessage(helloData);
}

json createChatMessage(const vector<string>& destinations, const string& iv, const vector<string>& symmKeys, const json& innerChat) {
    json chatData;
    chatData["type"] = "chat";
    chatData["destinations"] = destinations; // still needs implementation
    chatData["iv"] = iv; // still needs implementation
    chatData["symm_keys"] = symmKeys; // still needs implementation
    chatData["chat"] = innerChat;

    return createSignedMessage(chatData);
    
}

json innerChatMessage(const vector<string>& participants, const string& message) { // tracking of participants and whatnot needs implementation
    json chatInner;
    vector<string> base64Participants;
    for (const auto& participant : participants) {
        base64Participants.push_back(encodeFingerprint(participant));
    }
    chatInner["participants"] = base64Participants; // base64 encoded list of fingerprints of participants, starting with sender

    chatInner["message"] = message;


    return chatInner;
}

string encodeFingerprint(const string& fingerprint) { // helper function to encode a fingerprint in base64
    return macaron::Base64::Encode(fingerprint);
}


json publicChat(const string& fingerprint, const string& message) {
    json publicChatData;
    publicChatData["type"] = "public_chat";
    publicChatData["sender"] = encodeFingerprint(fingerprint); 
    publicChatData["message"] = message;
    return createSignedMessage(publicChatData); // unsure if we need public chat but got it here just in case
}

json createClientList(const vector<ServerInfo>& servers) { //still needs server tracking implementation?
    json clientList;
    clientList["type"] = "client_list";
    clientList["servers"] = json::array();
    
    for (const auto& server : servers) { // still needs implementation
        json serverInfo; 
        serverInfo["address"] = server.address; 
        serverInfo["clients"] = json::array();
        
        for (const auto& client : server.clients) {
            serverInfo["clients"].push_back(client); // add each client to the clients array
    }
    clientList["servers"].push_back(serverInfo);
    }
    return clientList;
}
int main() {




}

File: json.py
import json as js
import base64

m_counter = 0

class ClientMessageStructure:
    def __init__(self, message_type, data, counter, signature):
        self.type = message_type
        self.data = data
        self.counter = counter
        self.signature = signature

class ClientHelloData:
    def __init__(self, public_key):
        self.type = "hello"
        self.public_key = public_key

class ServerInfo:
    def __init__(self, address, clients):
        self.address = address
        self.clients = clients

class ClientList:
    def __init__(self, servers):
        self.type = "client_list"
        self.servers = servers

def create_signed_message(data):
    global m_counter
    message = {
        "type": "signed_data",
        "data": data,
        "counter": m_counter
    }

    message_str = js.dumps(data)
    signature_input = message_str + str(m_counter)
    signature = base64.b64encode(signature_input.encode()).decode

    message["signature"] = signature
    m_counter += 1
    return message

def hello_message(public_key):
    hello_data = ClientHelloData(public_key)
    return create_signed_message(hello_data.__dict__)

def create_chat_message(destinations, iv, symm_keys, inner_chat):
    chat_data = {
        "type": "chat",
        "destinations": destinations,
        "iv": iv,
        "symm_keys": symm_keys,
        "chat": inner_chat
    }
    return create_signed_message(chat_data)

def inner_chat_message(participants_signatures, message):
    chat_inner = {
        "participants": [encode_fingerprint(participant) for participant in participants_signatures],
        "message": message
    }
    return chat_inner

def encode_fingerprint(fingerprint):
    return base64.b64encode(fingerprint.encode()).decode()

def public_chat(fingerprint, message):
    public_chat_data = {
        "type": "public_chat",
        "sender": encode_fingerprint(fingerprint),
        "message": message
    }
    return create_signed_message(public_chat_data)

def create_client_list(servers):
    client_list = {
        "type": "client_list",
        "servers": []
    }
    
    for server in servers:
        server_info = {
            "address": server.address,
            "clients": server.clients
        }
        client_list["servers"].append(server_info)
    
    return client_list

def create_client_update(clients):
    client_update = {
        "type": "client_update",
        "clients": clients
    }
    return client_update

def create_client_update_request():
    return {"type": "client_update_request"}